flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,301 @@
+		implements ResultTypeQueryable<OUT>, ParallelSourceFunction<OUT>, StoppableFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSource.class);
+	protected DeserializationSchema<OUT> deserializationSchema;
+	protected SubscriberWrapper subscriberWrapper;
+
+	protected boolean running = true;
+	protected transient volatile SourceContext<OUT> sourceContext = null;
+
+	PubSubSource(DeserializationSchema<OUT> deserializationSchema, SubscriberWrapper subscriberWrapper) {
+		super(String.class);
+		this.deserializationSchema = deserializationSchema;
+		this.subscriberWrapper = subscriberWrapper;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		subscriberWrapper.initialize();
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesReceivedNotProcessed", subscriberWrapper::amountOfMessagesInBuffer);
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<AckReplyConsumer> ackReplyConsumers) {
+		ackReplyConsumers.forEach(AckReplyConsumer::ack);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		this.sourceContext = sourceContext;
+		subscriberWrapper.start();
+
+		while (subscriberWrapper.isRunning()) {
+			try {
+				Tuple2<PubsubMessage, AckReplyConsumer> newMessage = subscriberWrapper.take();
+				if (newMessage != null) {
+					processMessage(newMessage);
+				}
+			} catch (InterruptedException e) {
+				LOG.debug("Interrupted - stop or cancel called?");
+			}
+		}
+
+		nackOutstandingMessages();
+
+		LOG.debug("Waiting for PubSubSubscriber to terminate.");
+		subscriberWrapper.awaitTerminated();
+	}
+
+	void processMessage(Tuple2<PubsubMessage, AckReplyConsumer> input) {
+		PubsubMessage message = input.f0;
+		AckReplyConsumer ackReplyConsumer = input.f1;
+		if (sourceContext == null) {
+			ackReplyConsumer.nack();
+			return;
+		}
+
+		synchronized (sourceContext.getCheckpointLock()) {
+			boolean alreadyProcessed = !addId(message.getMessageId());
+			if (alreadyProcessed) {
+				return;
+			}
+
+			sessionIds.add(ackReplyConsumer);
+			OUT deserializedMessage = deserializeMessage(message);
+			sourceContext.collect(deserializedMessage);
+			if (deserializationSchema.isEndOfStream(deserializedMessage)) {
+				stop();
```

Comment:
It seems that `stop()` is stopping the message processing asynchronously. This behavior would not be in line with how the other connectors use the end of stream.
I think breaking out of the fetch loop would be the preferred approach (for consistency reasons)

Reply:
Would you mind if we keep it as it is currently?

The behavior should be consistent with the other sources as `isEndOfStream()` indicates the end of the stream has been reached so no new messages should arrive anyhow. And this way we don't make the code less readable by adding another way can break out of the fetch loop.